service/history/workflow/mutable_state_impl.go

```diff
@@ -806,6 +836,37 @@ func (e *MutableStateImpl) GetCronBackoffDuration() time.Duration {
 	return backoff.GetBackoffForNextSchedule(e.executionInfo.CronSchedule, executionTime, e.timeSource.Now())
 }
 
+// GetSearchAttributes returns the search attributes of the workflow execution. If the close execution visibility task
+// is incomplete, this method fetches the search attributes from the mutable state. Otherwise, it fetches the search
+// attributes from the visibility storage backend.
+func (e *MutableStateImpl) GetSearchAttributes(
+	ctx context.Context,
+	visibilityManager manager.VisibilityManager,
+) (*commonpb.SearchAttributes, error) {
+	executionInfo := e.GetExecutionInfo()
+	// if we have search attributes in mutable state, return it
+	if !executionInfo.GetCloseVisibilityTaskCompleted() {
+		return &commonpb.SearchAttributes{IndexedFields: executionInfo.SearchAttributes}, nil
+	}
+
+	// if we don't have search attributes in mutable state, try to get it from visibility store
+	executionState := e.GetExecutionState()
+	visResponse, err := visibilityManager.GetWorkflowExecution(
```

Comment:
Wondering if we should create a function to return both memo and search attributes to avoid making two calls to visibility storage.
cc: @yiminc @alexshtin